### PR TITLE
Fix eloquent models and collections

### PR DESCRIPTION
**Pull request by joshhanley**
This PR fixes issues with Eloquent Models and Collections when are assigned to public properties and then are deleted, throwing 404s and causing n+1 problems. See #6120, #6297, #6302, #6457, #6511, #6792, #6964.

As per a discussion with Caleb, we have decided to:
- Allow empty models to be assigned (gets rid of `Can't set model as property if it hasn't been persisted yet` error)
- If a model has been assigned to a property and then it has been deleted, then we just assign an empty model of the same class to the property (this stops a 404 issue) - we decided this was better than setting to `null`
- Add an Eloquent Collection synth that ensures Eloquent Collections are correctly rehydrated using a `whereIn()` query (this also stops a different 404 issue and also ensures that restoring collections don't cause an n+1 problem) 

I've merged the test from PR #6858 into this PR, so will close that one and #6272.

Hope this helps!

Edit: I have also added a warning box to the properties page regarding custom queries and Eloquent Collection and Model properties, under where Eloquent Collections and Models are listed for the first time, and added a link to the section at the bottom of the page which explains in more detail.

**I wonder if we should also mention in that warning box that any data stored on the Model/Collection that hasn't been persisted won't be preserved between requests?**

image

### CODE REVIEW

**Comment by calebporzio** [comment]
Legend @joshhanley 👊

File: src/Features/SupportModels/EloquentCollectionSynth.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    static function match($target)
+    {
+        return $target instanceof EloquentCollection;
+    }
+
+    function dehydrate(EloquentCollection $target, $dehydrateChild)
+    {
+        $class = $target::class;
+        $modelClass = $target->getQueueableClass();
```
Comment (QuentinGab) [inline src/Features/SupportModels/EloquentCollectionSynth.php:23]: `getQueueableClass` return null if the collection is empty. 
Then it throws error when calling`$modelAlias = (new $modelClass)->getMorphClass();`.

I'm not comfortable myself with Synths, maybe @joshhanley you have an idea on how to fix this ?

Comment (QuentinGab) [reply src/Features/SupportModels/EloquentCollectionSynth.php:23]: @calebporzio I always wonder if you would prefer to discuss newly introduced bug in the MR or in a new Discussion ?

Comment (QuentinGab) [reply src/Features/SupportModels/EloquentCollectionSynth.php:23]: I'm working on a fix myself

Comment (QuentinGab) [reply src/Features/SupportModels/EloquentCollectionSynth.php:23]: I submitted a PR #7017, tests are passing on my machine.

Comment (joshhanley) [reply src/Features/SupportModels/EloquentCollectionSynth.php:23]: For anyone else seeing this, fix has been released in v3.0.7.

**Comment by MarcoCazzaro** [comment]
Suggestion: put a giant banner in the [Upgrade Guide](https://livewire.laravel.com/docs/upgrading), please. I think it's a massive change, that usually run completely undetected. E.g. if like me the order in the database of my models is the same of their "display_order" field which I use in my orderBy method, everything _looks_ OK, but as soon as you change the order in the fields everything messes up and you have absolutely no clue.

**Comment by joshhanley** [comment]
@MarcoCazzaro we've just fixed that issue in PR #7359. So should be available in the next release.

**Pull request merged**